Add chart component and integrate with API for data visualization

adamski83 opened the pull request.

adamski83 committed:
Add chart component and integrate with API for data visualization

fengerkamil commented on src/AppRoutes.tsx:
```diff
@@ -18,6 +19,7 @@ export const AppRoutes = () => {
         <Route path={PATHS.qrScanner} element={<Invoices />} />
         <Route path="/edit/:id" element={<EditBox />} />
```
A ten path nie może być w zmiennej? To bez różnicy że ma parametr

fengerkamil commented on src/components/Charts/Chart.tsx:
```diff
@@ -0,0 +1,44 @@
+    }
+  }, [apiCall]);
+
+  const chartDataSet = {
+    labels: chartLabels,
+    datasets: [
+      {
+        label: t("charts.title"),
+        data: chartData,
+      },
+    ],
+  };
```
Opakowałbym to w useMemo które przelicza swoją wartość jedynie wtedy gdy zmieni się chartLabels lub chartData.

fengerkamil commented on src/components/Charts/Chart.tsx:
```diff
@@ -0,0 +1,44 @@
+
+  const [chartData, setChartData] = useState<number[]>([]);
+  const [chartLabels, setChartLabels] = useState<string[]>([]);
+  useEffect(() => {
+    if (apiCall?.data) {
+      const labels = apiCall?.data.map((box) => box.name || "Bez nazwy");
+      const values = apiCall?.data.map((box) => box.amount || 0);
+
+      setChartLabels(labels);
+      setChartData(values);
+    }
+  }, [apiCall]);
```
useEffecty zwykle definiujemy po zmiennych, funkcjach, ale przed returnem  lub innymi returnami jbc. Także do podmiany kolejność tego z chartDataSet

fengerkamil commented on src/components/Charts/Chart.tsx:
```diff
@@ -0,0 +1,44 @@
+  if (!chartData.length) return <div>Brak danych do wyświetlenia</div>;
+
+  return (
+    <div style={{ width: "100vw", height: "100vh" }}>
+      <Line ref={ref} data={chartDataSet} />
+    </div>
```
Nie używamy inline styli w żadnym wypadku

fengerkamil commented on src/components/sidebar/SidebarMenu.tsx:
```diff
@@ -37,6 +37,11 @@ export const SidebarMenu = ({ onClose }: any) => {
       icon: <SettingsApplicationsRoundedIcon onClick={onClose} />,
     },
+    {
+      to: "/Charts",
```
`/charts`

fengerkamil commented on src/components/sidebar/SidebarMenu.tsx:
```diff
@@ -37,6 +37,11 @@ export const SidebarMenu = ({ onClose }: any) => {
     },
+    {
+      to: "/Charts",
+      label: <>{t("sidebar.menu.charts")}</>,
```
Czy to opakowanie `<></>` jest absolutnie konieczne?

fengerkamil commented on src/components/sidebar/SidebarMenu.tsx:
```diff
@@ -37,6 +37,11 @@ export const SidebarMenu = ({ onClose }: any) => {
+    {
+      to: "/Charts",
+      label: <>{t("sidebar.menu.charts")}</>,
+      icon: <SettingsApplicationsRoundedIcon onClick={onClose} />,
```
W każdej ikonie dodajesz to samo, lepiej byłoby stworzyć funkcję powyżej menuItems który przyjmie ikonę i zwróci element JSX jako opakowaną ikonę buttonem, a ten button dostanie onClicka raz. Teraz to powielamy X razy

fengerkamil commented on src/utils/translations.ts:
```diff
@@ -55,6 +56,9 @@ interface TranslationsSchema {
       completed: string;
     };
   };
+  charts: {
+    title: string;
+  };
```
Nie musisz wgl dodawać typów do translacji. Wystarczy że obiekt z wartościami zdefiniujesz jako const i na samym końcu obiektu, przed `;` dodasz `as const` i tyle, VSC będzie sam podpowiadał właściwości wraz z wartościami

adamski83 committed:
Add chart type selection and improve chart layout in Chart component

adamski83 committed:
Refactor styles and types: remove unused CSS files, add new theme and layout types,moved file to correct directors

adamski83 committed:
Add Live component for real-time Arduino data visualization and update routing

fengerkamil commented on src/components/layouts/Live/Live.tsx:
```diff
@@ -0,0 +1,48 @@
+    error,
+    isLoading,
+  } = useArduinoData({
+    queryOptions: {
+      refetchInterval: 5000,
+      staleTime: 2000,
+      refetchOnWindowFocus: true,
+      refetchOnReconnect: true,
+    },
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ["arduinoData"] });
+      console.log("Box updated successfully");
+    },
```
Czy to invalidateQueries jest na pewno wymagane? Samo refetch interval załatwi Ci odswiezanie, po co ten onSuccess oprócz tego, że dodaje console loga?

fengerkamil commented on src/components/layouts/Live/Live.tsx:
```diff
@@ -0,0 +1,48 @@
+      console.log("Box updated successfully");
+    },
+  });
+  if (error) return <p>{error.message || String(error)}</p>;
```
Przygotuj się jeszcze na case aby wyswietlic swoj domyslny error, jakis string, bo nie zawsze error musi przyjść.
Poza tym `String(error)` to trochę ryzykowne podejscie, jak przyjdzie obiekt to go wyswietlisz?

fengerkamil commented on src/components/layouts/Live/Live.tsx:
```diff
@@ -0,0 +1,48 @@
+    queryOptions: {
+      refetchInterval: 5000,
+      staleTime: 2000,
+      refetchOnWindowFocus: true,
```
W sumie w tym całym obiekcie nowością jest tylko refetchInterval oraz staleTime (tu 2s. w globalnym configu 3s), resztajest juz w globalnym configu to nie wiem czy jest sens tutaj to powielać.

fengerkamil commented on src/components/layouts/Live/Live.tsx:
```diff
@@ -0,0 +1,48 @@
+    },
+  });
+  if (error) return <p>{error.message || String(error)}</p>;
+  if (isLoading) return <p>Loading...</p>;
```
Może dorobimy się jakiegoś fajnego loadera który się będzie kręcił i ładnie nam się wyświetli? :D

fengerkamil commented on src/components/layouts/Live/Live.tsx:
```diff
@@ -0,0 +1,48 @@
+      <h3>Dane z Arduino - aktualizacja co 5 sekund</h3>
+      <div className="data-card">
+        <p>
+          <strong>Ilość kartonów:</strong> {arduinoData?.data?.boxAmount || 0}
+        </p>
+        <p>
+          <strong>Odległość:</strong> {arduinoData?.data?.distance || 0} cm
+        </p>
+        <p>
+          <strong>Status:</strong>{" "}
+          {arduinoData?.connected ? "Połączono" : "Rozłączono"}
```
Przydałoby się:
- Te dane bardziej zdestrukturyzować przed returnem, bo powiemay arudinoData często
- Przetłumaczyć teksty
- Tak samo ze stanem połaczony/rozłączono

fengerkamil commented on src/components/layouts/Live/Live.tsx:
```diff
@@ -0,0 +1,48 @@
+        </p>
+        <p>
+          <strong>Ostatnia aktualizacja:</strong>{" "}
+          {new Date(arduinoData?.lastUpdate || Date.now()).toLocaleTimeString()}
```
To nie jest trochę oszustwo? :D jak nie ma danych z arudino, to ustalasz że ostatni update jest teraz? Lepiej wyświetlić jakąś informację że nie ma danych, lub że dane są nieświeże czy coś

fengerkamil commented on src/services/queries/getArduinoData.ts:
```diff
@@ -0,0 +1,37 @@
+export interface ArduinoData {
+  connected: boolean;
+  port: string;
+  error: string | null;
```
Tutaj bym popracował nad backendem, aby za wszelką cenę zawsze dostarczał stringa, jezeli arudino nic nie wygeneruje, to sam musisz o to zadbać. Error będący nullem będzie kiedyś nienamierzalny podczas błędów

fengerkamil commented on src/services/queries/getArduinoData.ts:
```diff
@@ -0,0 +1,37 @@
+    distance: number;
+    boxAmount: number;
+    maxDistance: number;
+    [key: string]: any;
```
Czy to aby na pewno potrzebne? Backend powinien zwracać tylko to, co potrzebujesz na Froncie, w rezultacie zwracasz to co znasz, więc możesz to o typować. Nie ma co zwracać nadmiaru opcji z których nie skorzystasz bo potem kusi właśnie do takiego typowania. Chyba ze powód jest inny?

fengerkamil commented on src/services/queries/getArduinoData.ts:
```diff
@@ -0,0 +1,37 @@
+
+export const useArduinoData = (options?: UseArduinoDataOptions) => {
+  return useQuery<ArduinoData>({
+    queryKey: ["arduinoData"],
```
Ten klucz powinien być w pliku tagów, poza tym fajnie byłoby zmienić trochę ich styl np `GET_ARUDINO_DATA` wtedy jak debugujesz w devtoolsach react query, to widzisz konkretną nazwę która wskazuje na typ pobieranych danych, teraz też jest okej, ale jak zaczniesz coś jeszcze pobierać z arudino to zacznie sie kombinowanie dziwnymi zmianami.

Nazywamy jasno, jednoznacznie

fengerkamil commented on src/services/queries/getArduinoData.ts:
```diff
@@ -0,0 +1,37 @@
+  return useQuery<ArduinoData>({
+    queryKey: ["arduinoData"],
+    queryFn: async () => {
+      const response = await axios.get("http://localhost:5001/arduino/data");
```
Ten API url (do ostatniego /) powinien być częścią .env, tak aby na potencjalnym prodzie można było odczytać inny url niż mamy lokalnie.

fengerkamil commented on src/services/queries/getArduinoData.ts:
```diff
@@ -0,0 +1,37 @@
+      const response = await axios.get("http://localhost:5001/arduino/data");
+      return response.data;
+    },
+    ...options?.queryOptions,
+    onSuccess: options?.onSuccess,
+    onError: options?.onError,
```
wystarczy `...options`

fengerkamil commented on src/urls/urls.ts:
```diff
@@ -7,6 +7,7 @@ type paths = {
   invoices: string;
   charts: string;
   edit: string;
+  lives: string;
```
te typy nie są potrzebne, jak usuniesz ten type, a w conscie ponizej dodasz na koncu za obiektem `as const` to bedziesz mial ten sam efekt a nie bedziesz musial typować tego raz a ponizej drugi raz definiować.

adamski83 committed:
feat: Remove unused types file, add LiveErrorHandler component, and implement MainLayout with responsive design

- Deleted the unused `types.ts` file.
- Added `LiveErrorHandler` component for error handling in live sessions, utilizing MUI for UI elements and i18next for translations.
- Implemented `MainLayout` component with responsive design, integrating mobile navigation and theme toggle functionality.
